format/luajit/luajit.go

```diff
@@ -0,0 +1,355 @@
+package luajit
+
+// dump   = header proto+ 0U
+// header = ESC 'L' 'J' versionB flagsU [namelenU nameB*]
+// proto  = lengthU pdata
+// pdata  = phead bcinsW* uvdataH* kgc* knum* [debugB*]
+// phead  = flagsB numparamsB framesizeB numuvB numkgcU numknU numbcU
+//          [debuglenU [firstlineU numlineU]]
+// kgc    = kgctypeU { ktab | (loU hiU) | (rloU rhiU iloU ihiU) | strB* }
+// knum   = intU0 | (loU1 hiU)
+// ktab   = narrayU nhashU karray* khash*
+// karray = ktabk
+// khash  = ktabk ktabk
+// ktabk  = ktabtypeU { intU | (loU hiU) | strB* }
+//
+// B = 8 bit, H = 16 bit, W = 32 bit, U = ULEB128 of W, U0/U1 = ULEB128 of W+1
+
+// see:
+//
+//	* http://scm.zoomquiet.top/data/20131216145900/index.html
+//	* https://github.com/LuaJIT/LuaJIT/blob/v2.1/src/lj_bcdump.h
+
+import (
+	"bytes"
+	"encoding/binary"
+
+	"golang.org/x/text/encoding"
+
+	"github.com/wader/fq/format"
+	"github.com/wader/fq/pkg/decode"
+	"github.com/wader/fq/pkg/interp"
+	"github.com/wader/fq/pkg/scalar"
+)
+
+func init() {
+	interp.RegisterFormat(
+		format.LuaJIT,
+		&decode.Format{
+			Description: "LuaJIT 2.0 bytecode dump",
+			DecodeFn:    LuaJITDecode,
+		})
+}
+
+// reinterpret an int as a float
+func u64tof64(u uint64) float64 {
+		var buf [8]byte
+
+		binary.BigEndian.PutUint64(buf[:], u)
+
+		var f float64
+		binary.Read(bytes.NewBuffer(buf[:]), binary.BigEndian, &f)
+
+		return f
+}
+
+type DumpInfo struct {
+	Strip bool
+	BigEndian bool
+}
+
+func LuaJITDecodeHeader(di *DumpInfo, d *decode.D) {
+	magic := d.FieldStr("magic", 3, encoding.Nop)
+
+	ok := bytes.Equal([]byte(magic), []byte{0x1b, 0x4c, 0x4a})
+	if !ok {
+		d.Errorf("invalid magic number")
+	}
```

Comment:
Think you can replace this with `d.FieldStr("magic", 3, d.AssertStr(".."))` or `d.FieldRawLen("magic", 3*8, d.AssertBitBuf([]byte{...}))` which will do the compare and also the magic will be shown as "valid"